libcperciva/util/sock.c

```diff
@@ -296,6 +296,65 @@ sock_resolve(const char * addr)
 	return (NULL);
 }
 
+/**
+ * sock_resolve_one(addr):
+ * Return a single sock_addr structure, or NULL if there are no addresses.
+ * Warn if there is more than one address, and return the first one.
+ */
+struct sock_addr *
+sock_resolve_one(const char * addr)
+{
+	struct sock_addr ** sas;
+	struct sock_addr * sa;
+
+	/* Resolve target address. */
+	if ((sas = sock_resolve(addr)) == NULL) {
+		warnp("Error resolving socket address: %s", addr);
+		goto err0;
+	}
+
+	/* Check that the array is not empty. */
+	if (sas[0] == NULL) {
+		warn0("No addresses found for %s", addr);
+		goto err1;
+	}
+
+	/* If there's more than one address, give a warning. */
+	if (sas[1] != NULL)
+		warn0("Using the first of multiple addresses found for %s",
+		    addr);
+
+	/* Allocate a structure. */
```

Comment:
Err... we have `sock_addr_dup` if you want to do it that way.  But why not just free the bits we don't want?

```
sa = sas[0];
for (p = &sas[1]; *p != NULL; p++)
    sock_addr_free(*p);
free(sas);
```

Reply:
Huh, I never noticed `sock_util.h`.

I agree that avoiding the malloc is a cleaner solution.